jupyterlab/handlers/yjs_echo_ws.py

```diff
@@ -25,170 +25,163 @@
+    def get_room(self, path: str) -> JupyterRoom:
+        file_type, file_path = path.split(":", 1)
+        room = self.rooms.get(file_path, JupyterRoom(file_type))
+        self.rooms[file_path] = room
+        return room
 
-class YjsRoom:
-    def __init__(self, type):
-        self.type = type
-        self.clients = {}
-        self.content = bytes([])
-        self.ydoc = YDOCS.get(type, YFILE)()
 
-    def get_source(self):
-        return self.ydoc.source
+WEBSOCKET_SERVER = JupyterWebsocketServer(has_internal_ydoc=True, auto_clean_rooms=False)
 
 
 class YjsEchoWebSocket(WebSocketHandler, JupyterHandler):
 
+    saving_document: Optional[asyncio.Task]
+
     # Override max_message size to 1GB
     @property
     def max_message_size(self):
         return 1024 * 1024 * 1024
 
+    def __aiter__(self):
+        # needed to be compatible with WebsocketServer (async for message in websocket)
+        return self
+
+    async def __anext__(self):
+        # needed to be compatible with WebsocketServer (async for message in websocket)
+        message = await self._message_queue.get()
+        if not message:
+            raise StopAsyncIteration()
+        return message
+
     async def get(self, *args, **kwargs):
         if self.get_current_user() is None:
             self.log.warning("Couldn't authenticate WebSocket connection")
             raise web.HTTPError(403)
         return await super().get(*args, **kwargs)
 
-    def open(self, type_path):
-        # print("[YJSEchoWS]: open", type_path)
-        type, path = type_path.split(":", 1)
-        self.id = str(uuid.uuid4())
-        self.room_id = path
-        room = ROOMS.get(self.room_id)
-        if room is None:
-            room = YjsRoom(type)
-            ROOMS[self.room_id] = room
-        room.clients[self.id] = (IOLoop.current(), self.hook_send_message, self)
-        # Send SyncStep1 message (based on y-protocols)
-        self.write_message(bytes([0, 0, 1, 0]), binary=True)
+    async def open(self, path):
+        self.path = path  # needed to be compatible with WebsocketServer (websocket.path)
+        self._message_queue = asyncio.Queue()
+        self.file_type, self.file_path = path.split(":", 1)
+        self.saving_document = None
+        self.room = WEBSOCKET_SERVER.get_room(path)
+        asyncio.create_task(WEBSOCKET_SERVER.serve(self))
+
+        # cancel the deletion of the room if it was scheduled
+        if self.room.cleaner is not None:
+            self.room.cleaner.cancel()
+
+        if not self.room.ydoc.initialized.is_set():
+            model = await ensure_async(
+                self.contents_manager.get(self.file_path, type=self.file_type)
+            )
+            self.last_modified = model["last_modified"]
+            # check again if initialized, because loading the file can be async
+            if not self.room.ydoc.initialized.is_set():
+                self.room.document.source = model["content"]
```

Comment:
This seems to handle all documents as a text document, is that correct?

Reply:
Not exactly, for a generic file the returned source is a text indeed, but for a notebook it is a JSON object (as a Python dictionary).